hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/DefaultProviderSelector.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.security.provider;
+
+import java.util.Map;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public class DefaultProviderSelector implements AuthenticationProviderSelector {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultProviderSelector.class);
+
+  Configuration conf;
+  SimpleSaslClientAuthenticationProvider simpleAuth = null;
+  GssSaslClientAuthenticationProvider krbAuth = null;
+  DigestSaslClientAuthenticationProvider digestAuth = null;
+  Map<Byte,SaslClientAuthenticationProvider> providers;
+
+  @Override
+  public void configure(Configuration conf, Map<Byte,SaslClientAuthenticationProvider> providers) {
+    this.conf = conf;
+    this.providers = providers;
+    for (SaslClientAuthenticationProvider provider : providers.values()) {
+      if (provider instanceof SimpleSaslClientAuthenticationProvider) {
+        if (simpleAuth != null) {
+          LOG.warn("Ignoring duplicate SimpleSaslClientAuthenticationProvider: previous={},"
```

Comment:
Good point -- we really don't need to hold onto the `providers`.

Thinking some more, the expectation is that `configure` is only ever called once. We can catch and error out when it's called multiple times.

The logic I added here was if two implementations happened to not be uniquely identified from each other (e.g. a coding bug). If that happens, we should just bail out because that's a bad condition (we could be using an unexpected version of some provider).